### PR TITLE
Detune to stop ground oscillation

### PR DESCRIPTION
**Pull request by lthall**
This pr provides the ability to reduce the gains of roll, pitch, and yaw to remove ground oscillation while landed. Each axis has a gain reduction variable that is set when the aircraft is landed. The gain returns to normal in 2 x ATC_INPUT_TC.

This is a common problem on aircraft with flexible landing gear. It is often seen as a yaw oscillation during spool up on more solidly built aircraft also.

### CODE REVIEW

File: ArduCopter/mode_stabilize.cpp [hunk 1]
```diff
@@ -62,6 +62,9 @@ void ModeStabilize::run()
         break;
     }
 
+    // reduce gains when landed;
+    attitude_control->landed_gain_reduction(copter.ap.land_complete);
```
Comment (andyp1per) [inline ArduCopter/mode_stabilize.cpp:66]: Only stabilize?

File: libraries/AC_AttitudeControl/AC_AttitudeControl.cpp [hunk 2]
```diff
@@ -150,6 +150,27 @@ const AP_Param::GroupInfo AC_AttitudeControl::var_info[] = {
     // @User: Standard
     AP_GROUPINFO("INPUT_TC", 20, AC_AttitudeControl, _input_tc, AC_ATTITUDE_CONTROL_INPUT_TC_DEFAULT),
 
+    // @Param: LAND_R_MULT
+    // @DisplayName: Roll gain multiplier used when landed
```
Comment (andyp1per) [inline libraries/AC_AttitudeControl/AC_AttitudeControl.cpp:154]: I know a multiplier makes sense mathematically, but no-one is ever going to set it above 1 and so its really a divider. I would prefer a name that conveys to the use a bit more what it does.

Comment (MichelleRos) [reply libraries/AC_AttitudeControl/AC_AttitudeControl.cpp:154]: That would be a bit confusing. Unless we also invert the parameter...

Comment (lthall) [reply libraries/AC_AttitudeControl/AC_AttitudeControl.cpp:154]: I think people understand that multiplying by a number between 0 and 1 will reduce the gains.

**Comment by rmackay9** [comment]
@lthall I'm sure this will be popular.  Do we need to have both Roll and Pitch parameters?  I wonder if we can imagine a frame where we'd want them to be different.

**Comment by bnsgeyer** [comment]
> @lthall I'm sure this will be popular. Do we need to have both Roll and Pitch parameters? I wonder if we can imagine a frame where we'd want them to be different.

I think for heli there could be significantly different gains between pitch and roll.  But I don't know if the same backoff factor could work for both axes.

File: ArduCopter/mode_stabilize.cpp (hunk 1 above)
Comment (bnsgeyer) [reply ArduCopter/mode_stabilize.cpp:66]: @lthall do you intend to have this work in all modes at some point. I'm curious like @andyp1per as to why this is just stabilize

Comment (lthall) [reply ArduCopter/mode_stabilize.cpp:66]: Yes, this is just an example of how I need to do it right now. I was looking for suggestions on how I could do it better.

**Comment by lthall** [comment]
I considered this but I believe it would be a mistake to not separate them. We don't want to compromise control authority unnecessarily.  There are numerous aircraft that have significantly different ground interaction modes between all three axis.

File: ArduCopter/mode.cpp [hunk 3]
```diff
@@ -420,6 +420,7 @@ bool Copter::set_mode(const uint8_t new_mode, const ModeReason reason)
 void Copter::update_flight_mode()
 {
     surface_tracking.invalidate_for_logging();  // invalidate surface tracking alt, flight mode will set to true if used
+    attitude_control->landed_gain_reduction(copter.ap.land_complete); // Adjust gains when landed to attenuate ground oscillation
```
Comment (lthall) [inline ArduCopter/mode.cpp:423]: I have called this feature here to capture all flight modes in a single line. It seems like a good fit but I would like to get others feedback.

Comment (andyp1per) [reply ArduCopter/mode.cpp:423]: is land_complete all that you need?

**Review by andyp1per — APPROVED** [review]

File: ArduCopter/mode.cpp (hunk 3 above)
Comment (lthall) [reply ArduCopter/mode.cpp:423]: Yes, this is a reasonably conservative test.

**Review by IamPete1 — APPROVED** [review]
Looks good. Just to confirm, it does should up in the `ATSC` log message?

**Comment by lthall** [comment]
@IamPete1 Sorry mate, I didn't understand what you were asking.

**Comment by IamPete1** [comment]
> @IamPete1 Sorry mate, I didn't understand what you were asking.

The `ATSC`  message contains the used values of the angle P scale and the rate PD scale if all ones. The logging should "just work" with your changes, but it would be worth double checking just so we can track what is going on. You should see them slew up and down, Once they get to all ones the log should stop.

**Comment by lthall** [comment]
image

File: libraries/AC_AttitudeControl/AC_AttitudeControl.cpp [hunk 4]
```diff
@@ -204,6 +225,21 @@ void AC_AttitudeControl::reset_rate_controller_I_terms_smoothly()
     get_rate_yaw_pid().relax_integrator(0.0, _dt, AC_ATTITUDE_RATE_RELAX_TC);
 }
 
+// Reduce attitude control gains while landed to stop ground resonance
+void AC_AttitudeControl::landed_gain_reduction(bool landed)
+{
+    // use 2.0 x tc to match the response time to 86% commanded
+    const float spool_step = _dt / (2.0 * _input_tc);
```
Comment (rmackay9) [inline libraries/AC_AttitudeControl/AC_AttitudeControl.cpp:232]: I wonder if we should catch the possible divide by zero here

File: libraries/AC_AttitudeControl/AC_AttitudeControl.cpp [hunk 5]
```diff
@@ -150,6 +150,27 @@ const AP_Param::GroupInfo AC_AttitudeControl::var_info[] = {
...
 
+    // @Param: LAND_R_MULT
+    // @DisplayName: Roll gain multiplier used when landed
+    // @Description: Roll gain multiplier used when landed
+    // @Range: 0 1
+    // @User: Advanced
+    AP_GROUPINFO("LAND_R_MULT", 21, AC_AttitudeControl, _land_roll_mult, 1.0),
+
+    // @Param: LAND_P_MULT
+    // @DisplayName: Pitch gain multiplier used when landed
```
Comment (rmackay9) [inline libraries/AC_AttitudeControl/AC_AttitudeControl.cpp:161]: we should consider putting a Land at the front of the Display name

File: libraries/AC_AttitudeControl/AC_AttitudeControl.cpp [hunk 6]
```diff
@@ -150,6 +150,27 @@ const AP_Param::GroupInfo AC_AttitudeControl::var_info[] = {
     // @User: Standard
     AP_GROUPINFO("INPUT_TC", 20, AC_AttitudeControl, _input_tc, AC_ATTITUDE_CONTROL_INPUT_TC_DEFAULT),
 
+    // @Param: LAND_R_MULT
+    // @DisplayName: Roll gain multiplier used when landed
+    // @Description: Roll gain multiplier used when landed
```
Comment (rmackay9) [inline libraries/AC_AttitudeControl/AC_AttitudeControl.cpp:155]: We should give some instructions on what lower vs higher values do.  E.g. "lower values result in lower gains during landing, higher values result in higher gains during landing"

File: libraries/AC_AttitudeControl/AC_AttitudeControl.cpp [hunk 7]
```diff
@@ -150,6 +150,27 @@ const AP_Param::GroupInfo AC_AttitudeControl::var_info[] = {
...
+    // @Range: 0 1
```
Comment (rmackay9) [inline libraries/AC_AttitudeControl/AC_AttitudeControl.cpp:156]: Is the lower range really 0?

File: libraries/AC_AttitudeControl/AC_AttitudeControl.cpp (hunk 4 above)
Comment (lthall) [reply libraries/AC_AttitudeControl/AC_AttitudeControl.cpp:232]: I am not sure if I like what I did here.

File: libraries/AC_AttitudeControl/AC_AttitudeControl.cpp (hunk 5 above)
Comment (lthall) [reply libraries/AC_AttitudeControl/AC_AttitudeControl.cpp:161]: Maybe 
"Landed pitch gain multiplier"
"Landed pitch gain multiplier, only active when landed" (seems to say it twice though)

File: libraries/AC_AttitudeControl/AC_AttitudeControl.cpp (hunk 7 above)
Comment (lthall) [reply libraries/AC_AttitudeControl/AC_AttitudeControl.cpp:156]: Changing to 0.25 to 1.0

File: libraries/AC_AttitudeControl/AC_AttitudeControl.cpp (hunk 6 above)
Comment (lthall) [reply libraries/AC_AttitudeControl/AC_AttitudeControl.cpp:155]: Does this work?
    // @Description: Roll gain multiplier used when landed. A factor of 1.0 means no reduction in gain while landed. Reduce this factor to reduce ground oscitation in the roll axis.

**Review by rmackay9 — APPROVED** [review]
LGTM, thanks for the changes

**Pull request merged**